docs/src/main/asciidoc/building-my-first-extension-guide.adoc

```diff
@@ -0,0 +1,656 @@
+////
+This guide is maintained in the main Quarkus repository
+and pull requests should be submitted there:
+https://github.com/quarkusio/quarkus/tree/master/docs/src/main/asciidoc
+////
+= Quarkus - Building my first extension
+
+include::./attributes.adoc[]
+
+https://quarkus.io/extensions/[Quarkus extensions] enhance your application just as projects dependencies do.
+The role of the extensions is about leveraging Quarkus paradigms to integrate seamlessly a library into Quarkus.
+This is how you can use your battle-tested ecosystem and take advantage of Quarkus performance and native compilation.
+
+During this guide you understand how to develop the *Sample Greeting Extension*.
+The extension will expose a customizable HTTP endpoint which simply greet the visitor.
+
+== Prerequisites
+
+To complete this guide, you need:
+
+* less than 30 minutes
+* https://github.com/quarkusio/quarkus[Quarkus sources]
+* an IDE
+* JDK 1.8+ installed with JAVA_HOME configured appropriately
+* Apache Maven 3.5.3+
+
+== Basic Concepts
+
+First things first, we will need to start with some basic concepts.
+
+* JVM mode vs Native mode
+  ** Quarkus is first and foremost a Java framework, that means you can develop, package and run classic JAR application, that's what we call *JVM mode*.
+  ** Thanks to https://graalvm.org/[GraalVM] you can compile your Java application into machine specific code (like you do in Go or C++) and that's what we call *Native mode*.
+
+The operation of compiling Java bytecode into a native system-specific machine code is named *Ahead of Time Compilation* (aka AoT).
+
+* build time vs runtime
+  ** The build time corresponds to all the actions you apply to your Java source files to convert them into something runnable (class files, jar/war, native images). Usually this stage is composed of the compilation, annotation processing, bytecode generation, etc. At this point, everything is under developer's scope and control.
```

Comment:
I'd add reading XML config file, scanning classes for specific annotations etc. To make it a bit more concrete than "annotation processing" which has an overloaded term.

Reply:
🤔 The purpose of this section is to explain Build time and Runtime in a Java world (pre-Quarkus era). In this world,  XML reading or class scanning are usually done at *Runtime*.

But.... for people familiar with Quarkus, this paragraph could be misleading. Maybe I should rename the section *build time vs runtime with traditional Java frameworks"

Reply:
ah I read too quickly and got confused indeed.